src/main/java/mops/businesslogic/FileServiceImpl.java

```
@@ -0,0 +1,213 @@
+import java.io.InputStream;
+import java.util.List;
+import java.util.Set;
+
+@Service
+@AllArgsConstructor
+@Slf4j
+public class FileServiceImpl implements FileService {
+
+    /**
+     * Service for permission checks.
+     */
+    private final DirectoryService directoryService;
+    /**
+     * Service for saving and retrieving file meta data.
+     */
+    private final FileInfoService fileInfoService;
+    /**
+     * File content repository.
+     */
+    private final FileRepository fileRepository;
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    @SuppressWarnings("PMD.LawOfDemeter")
+    @Transactional(rollbackFor = MopsException.class)
+    public void saveFile(Account account, long dirId, MultipartFile multipartFile,
+                         Set<String> tags) throws MopsException {
+        UserPermission userPermission = directoryService.getPermissionsOfUser(account, dirId);
+
+        if (!userPermission.isWrite()) {
+            log.error("User {} tried to save a file without write permission.",
+                    account.getName(),
+                    new WriteAccessPermissionException("No write permission")
+            );
+        }
+        //no Law of demeter violation
+        FileInfo meta = FileInfo.builder()
+                .from(multipartFile)
+                .directory(dirId)
+                .owner(account.getName())
+                .tags(tags)
+                .build();
+
+        try {
+            FileInfo fileInfo = fileInfoService.saveFileInfo(meta);
+            fileRepository.saveFile(multipartFile, fileInfo.getId());
+        } catch (MopsException e) {
+            TransactionAspectSupport.currentTransactionStatus().setRollbackOnly();
+            log.error("Error while saving file {} by user {}. Error: {}",
+                    meta.getName(),
+                    account.getName(),
+                    e.getMessage(),
+                    new MopsException("Error while saving", e)
```

you need to throw this exception instead of giving it to the logger.